python/semantic_kernel/planning/action_planner/action_planner.py

```diff
@@ -0,0 +1,297 @@
+import json
+import os
+from logging import Logger
+from textwrap import dedent
+from typing import Optional
+
+import regex
+
+from semantic_kernel import Kernel
+from semantic_kernel.orchestration.sk_context import SKContext
+from semantic_kernel.orchestration.sk_function_base import SKFunctionBase
+from semantic_kernel.planning import Plan
+from semantic_kernel.skill_definition import sk_function, sk_function_context_parameter
+from semantic_kernel.skill_definition.function_view import FunctionView
+from semantic_kernel.skill_definition.parameter_view import ParameterView
+from semantic_kernel.utils.null_logger import NullLogger
+
+
+class ActionPlanner:
+    """
+    Action Planner allows to select one function out of many, to achieve a given goal.
+    The planner implements the Intent Detection pattern, uses the functions registered
+    in the kernel to see if there's a relevant one, providing instructions to call the
+    function and the rationale used to select it. The planner can also return
+    "no function" if nothing relevant is available.
+    """
+
+    _stop_sequence: str = "#END-OF-PLAN"
+    _skill_name: str = "this"
+
+    _planner_function: SKFunctionBase
+
+    _kernel: Kernel
+    _prompt_template: str
+    _logger: Logger
+
+    def __init__(
+        self,
+        kernel: Kernel,
+        prompt: Optional[str] = None,
+        logger: Optional[Logger] = None,
+    ) -> None:
+        if kernel is None:
+            raise ValueError("Kernel cannot be `None`.")
+
+        self._logger = logger if logger else NullLogger()
+
+        __cur_dir = os.path.dirname(os.path.abspath(__file__))
+        __prompt_file = os.path.join(__cur_dir, "skprompt.txt")
+
+        self._prompt_template = prompt if prompt else open(__prompt_file, "r").read()
+
+        self._planner_function = kernel.create_semantic_function(
+            skill_name=self._skill_name,
+            prompt_template=self._prompt_template,
+            max_tokens=1024,
+            stop_sequences=[self._stop_sequence],
+        )
+        kernel.import_skill(self, self._skill_name)
+
+        self._kernel = kernel
+        self._context = kernel.create_new_context()
+
+    async def create_plan_async(self, goal: str) -> Plan:
+        """
+        :param goal: The input to the planner based on which the plan is made
+        :return: a Plan object
+        """
+
+        if goal is None:
+            raise ValueError("Goal cannot be `None`.")
+
+        self._logger.info(f"Finding the best function for achieving the goal: {goal}")
+
+        self._context.variables.update(goal)
+
+        generated_plan_raw = await self._planner_function.invoke_async(
+            context=self._context
+        )
+        generated_plan_raw_str = str(generated_plan_raw)
+
+        self._logger.info(f"Plan generated by ActionPlanner:\n{generated_plan_raw_str}")
+
+        # Ignore additional text around JSON recursively
+        json_regex = r"\{(?:[^{}]|(?R))*\}"
```

Comment:
There are a number of scenarios where AI will not generate a valid plan: perhaps it doesn't think the current list of functions are appropriate, service throttling, etc.

I think there should be a check here on whether a plan was actually found.